### PR TITLE
fix #2517 Don't compare if p_pref is empty

### LINKED ISSUE #2517: `contains()` with an empty string argument leaves Redis hanging at CPU 100% indefinitely
Here's a reproduction scenario, solely through the CLI:

```
127.0.0.1:6379> FLUSHALL
OK
127.0.0.1:6379> FT.CREATE some_idx PREFIX 1 "any_prefix" SCHEMA some_field TEXT
OK
127.0.0.1:6379> FT.AGGREGATE some_idx * FILTER "contains('any_string', '')"
1) (integer) 0
127.0.0.1:6379> HSET any_prefix:1 some_field 'value'
(integer) 1
127.0.0.1:6379> FT.AGGREGATE some_idx * FILTER "contains('any_string', '')"
```

At this point, Redis is left in a state where its CPU usage hovers around 100%, and further requests to it are blocked. I've seen it stay in this state for more than 4 hours, so I assume it's indefinite.

### PR DESCRIPTION
**Pull request by gkorland**
related to #1776

### CODE REVIEW

File: src/aggregate/functions/string.c [hunk 1]
```diff
@@ -340,13 +340,16 @@ static int stringfunc_contains(ExprEval *ctx, RSValue *result, RSValue **argv, s
...
+  size_t p_pref_size;
+  const char *p_pref = (char *)RSValue_StringPtrLen(pref, &p_pref_size);
   result->t = RSValue_Number;
 
   size_t num = 0;
-  while ((p_str = strstr(p_str, p_pref)) != NULL) {
-    num++;
-    p_str++;
+  if(p_pref_size > 0) {
+    while ((p_str = strstr(p_str, p_pref)) != NULL) {
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/aggregate/functions/string.c:349]: ignore

[the remark this answers is not in the document]
Comment (gkorland) [reply src/aggregate/functions/string.c:349]: ignore

File: src/aggregate/functions/string.c [hunk 2]
```diff
@@ -340,13 +340,16 @@ static int stringfunc_contains(ExprEval *ctx, RSValue *result, RSValue **argv, s
   RSValue *pref = RSValue_Dereference(argv[1]);
 
   char *p_str = (char *)RSValue_StringPtrLen(str, NULL);
-  const char *p_pref = (char *)RSValue_StringPtrLen(pref, NULL);
+  size_t p_pref_size;
+  const char *p_pref = (char *)RSValue_StringPtrLen(pref, &p_pref_size);
   result->t = RSValue_Number;
 
   size_t num = 0;
```
Comment (oshadmi) [inline src/aggregate/functions/string.c:347]: In python (3.8.2) the number of occurrences of an empty string is len+1
e.g., `print("ABC".replace("", "x"))` returns `xAxBxCx`

File: src/aggregate/functions/string.c [hunk 3]
```diff
@@ -339,14 +339,20 @@ static int stringfunc_contains(ExprEval *ctx, RSValue *result, RSValue **argv, s
...
-  while ((p_str = strstr(p_str, p_pref)) != NULL) {
-    num++;
-    p_str++;
+  if(p_pref_size > 0) {
+    while ((p_str = strstr(p_str, p_pref)) != NULL) {
+      num++;
+      p_str++;
+    }
+  } else if (p_pref_size == 0) {
+    num = p_str_size + 1;
```
Comment (oshadmi) [inline src/aggregate/functions/string.c:355]: Only declare it before
`size_t num;`
And
```suggestion
  if(p_pref_size > 0) {
    num = 0;
    while ((p_str = strstr(p_str, p_pref)) != NULL) {
      num++;
      p_str++;
    }
  } else {
    num = p_str_size + 1;
```

**Review by oshadmi — COMMENTED** [review]
Small comment

**Review by oshadmi — APPROVED** [review]
👍🏼 

Regarding documentation, it says:
`Return the number of occurrences of s2 in s1, 0 otherwise.`

So it can remain as-is
But we can also specifically mention an empty s2 will return "length of s1 + 1"
So users avoid unpleasant surprises such as infinite loops.

**Review by oshadmi — APPROVED** [review]

**Pull request merged**